load-testing/load_test.py

```diff
@@ -0,0 +1,140 @@
+import argparse
+import json
+import multiprocessing
+import os
+import sys
+import time
+from datetime import datetime
+
+from config import (
+    DETECTOR_NAME,
+    DETECTOR_QUERY,
+    ENDPOINT_URL,
+    GROUNDLIGHT_API_TOKEN,
+    IMAGE_PATH,
+    LOG_FILE,
+    REQUESTS_PER_SECOND,
+    TIME_BETWEEN_RAMP,
+)
+from groundlight import Detector, Groundlight
+from parse_load_test_logs import show_load_test_results
+
+if ENDPOINT_URL == "":
+    raise ValueError("ENDPOINT_URL cannot be an empty string.")
+
+if GROUNDLIGHT_API_TOKEN == "":
+    GROUNDLIGHT_API_TOKEN = os.environ.get("GROUNDLIGHT_API_TOKEN")
+    if GROUNDLIGHT_API_TOKEN is None:
+        raise ValueError("The GROUNDLIGHT_API_TOKEN must be set in config.py or as an environment variable.")
+
+
+def send_image_requests(  # noqa: PLR0913
+    process_id: int, detector: Detector, gl_client: Groundlight, num_requests_per_second: float, duration, log_file: str
+):
+    """Sends image requests to a Groundlight endpoint for a specified duration and logs results."""
+    os.makedirs(os.path.dirname(log_file), exist_ok=True)
+
+    # Prevent errors from appearing in terminal
+    sys.stderr = open(os.devnull, "w")
+
+    start_time = time.time()
+    request_number = 1
+
+    while time.time() - start_time < duration:
+        request_start_time = time.time()
+
+        log_data = {
+            "asctime": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
+            "worker_number": process_id,
+            "request_number": request_number,
+        }
+
+        try:
+            gl_client.ask_ml(detector=detector, image=IMAGE_PATH, wait=1)
+            log_data.update({"latency": round(time.time() - request_start_time, 4), "success": True})
+        except Exception as e:
+            log_data.update({"latency": round(time.time() - request_start_time, 4), "success": False, "error": str(e)})
+
+        with open(log_file, "a") as log:
+            log.write(json.dumps(log_data) + "\n")
+
+        request_number += 1
+
+        time.sleep(max(0, (1 / num_requests_per_second) - (time.time() - request_start_time)))
+
+
+def ramp_up_processes(max_processes, step_size, requests_per_second, detector, gl_client, custom_ramp: bool = False):  # noqa: PLR0913
+    """Ramps up the number of client processes over time."""
+    if os.path.exists(LOG_FILE):
+        os.remove(LOG_FILE)
+    os.makedirs(os.path.dirname(LOG_FILE), exist_ok=True)
+
+    if custom_ramp:
+        ramp_steps = [1, 2, 3, 4, 5, 10, 15, 20, 30, 40, 50, 60]
+        print(f"Using custom ramp schedule: {ramp_steps} with {TIME_BETWEEN_RAMP} seconds between each step.")
```

Comment:
This seems like a pre-set ramp rather than something custom.

Reply:
Ah, I see there is supposed to be a CLI option for custom_ramp. Perhaps that should take in the custom ramp schedule directly?

Reply:
I was intending this to be a preset option, since it's a pretty nice schedule that we'll probably reuse in the future. The CLI option just tells the function to use this preset. I should either rename these or actually allow a custom schedule to be inputted, though.